File: app/src/main/java/com/bykea/pk/partner/ui/withdraw/WithdrawalViewModel.java
package com.bykea.pk.partner.ui.withdraw;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.bykea.pk.partner.dal.source.remote.data.PersonalInfoData;
import com.bykea.pk.partner.dal.source.remote.data.WithdrawPaymentMethod;
import com.bykea.pk.partner.dal.source.withdraw.WithdrawRepository;
import com.bykea.pk.partner.models.data.SettingsData;
import com.bykea.pk.partner.ui.helpers.AppPreferences;

import java.util.List;

/**
 * This is the view model class of {WithdrawalActivity}
 *
 * @author Arsal Imam
 */
public class WithdrawalViewModel extends ViewModel {

    private MutableLiveData<List<WithdrawPaymentMethod>> mAvailablePaymentMethods;
    private WithdrawRepository withdrawRepository;
    private MutableLiveData<Boolean> _showConfirmationDialog, _onWithdrawCompleted, _showLoader;
    private MutableLiveData<Integer> _balanceInt;
    private MutableLiveData<String> _errorMessage;
    private MutableLiveData<PersonalInfoData> driverProfile;
    private WithdrawPaymentMethod selectedPaymentMethod;


    /**
     * Constructor of this viewModel
     * @param withdrawRepository instance to get and update data
     */
    public WithdrawalViewModel(WithdrawRepository withdrawRepository) {
        this.withdrawRepository = withdrawRepository;
        _showLoader = new MutableLiveData<>();
        _balanceInt = new MutableLiveData<>();
        _errorMessage = new MutableLiveData<>();
        _onWithdrawCompleted = new MutableLiveData<>();
        _showConfirmationDialog = new MutableLiveData<>();
    }

    /**
     * Load user (driver) profile from repository
     */
    private void loadUserProfile() {
        withdrawRepository.getDriverProfile(new WithdrawRepository.LoadWithdrawalCallback<PersonalInfoData>() {

            @Override
            public void onDataLoaded(PersonalInfoData data) {
                driverProfile.setValue(data);
            }

            @Override
            public void onDataNotAvailable(String errorMsg) {
                _showLoader.setValue(false);
            }
        });
    }

    /**
     * Load all payment methods from repository
     */
    private void loadWithdrawalMethods() {
        _showLoader.setValue(true);
        withdrawRepository.getAllPaymentMethods(new WithdrawRepository.LoadWithdrawalCallback<List<WithdrawPaymentMethod>>() {

            @Override
            public void onDataLoaded(List<WithdrawPaymentMethod> data) {
                _showLoader.setValue(false);
                data.get(0).setSelected(true);
                selectedPaymentMethod = data.get(0);
                mAvailablePaymentMethods.setValue(data);
            }

            @Override
            public void onDataNotAvailable(String errorMsg) {
                _showLoader.setValue(false);
                _errorMessage.setValue("معذرت فی الحال آپ کی ردخواست پر عمل نہیں کیا جا سکتا");
            }
        });
    }

    /**
     * Executes withdrawal PUT API from repository to perform withdrawal operation.
     */
    public void confirmWithdraw() {
        Integer enteredBalance = _balanceInt.getValue();
        _showLoader.setValue(true);
        withdrawRepository.performWithdraw(enteredBalance, selectedPaymentMethod.getCode(), new WithdrawRepository.LoadWithdrawalCallback<Boolean>() {

            @Override
            public void onDataLoaded(Boolean data) {
                _showLoader.setValue(false);
                _onWithdrawCompleted.setValue(true);
            }

            @Override
            public void onDataNotAvailable(String errorMsg) {
                _showLoader.setValue(false);
                _errorMessage.setValue("معذرت فی الحال آپ کی ردخواست پر عمل نہیں کیا جا سکتا");
            }
        });
    }

    /**
     * Binded with button component (in xml),
     * @param amt amount entered by the user
     */
    public void onSubmitClicked(String amt) {
        try {
            int amount = Integer.valueOf(amt);
            String errorMsg = validateContent(amount);
            _errorMessage.setValue(errorMsg);
            if (errorMsg == null) {
                _balanceInt.setValue(amount);
                _showConfirmationDialog.setValue(true);
            }
        } catch (NumberFormatException e) {
            e.printStackTrace();
        }
    }

    /**
     * Return the cnic number from driver profile
     * @return cnic number
     */
    public String getDriverCnicNumber() {
        return driverProfile.getValue() == null ?
                "" : driverProfile.getValue().getCnic();
    }

    /**
     * Performing validation on form submit
     * @param amount to withdraw
     *
     * @return <b>null</b> if validation done successfully, otherwise returns error message
     */
    public String validateContent(int amount) {
        SettingsData settingsData = AppPreferences.getSettings();

        double maxValue = settingsData.getSettings().getWithdrawPartnerMaxLimit(),
                minValue = settingsData.getSettings().getWithdrawPartnerMinLimit(),
                accountBalance = driverProfile.getValue().getWallet();
        if (amount < minValue) {
            return String.format("درج کردہ رقم" + " %s " + "سے کم نہیں ہوسکتی", Math.round(minValue));
        }
        if (amount > maxValue) {
            return String.format("ایک وقت میں" + " %s " + "سے زیادہ کی رقم نہیں نکالی جاسکتی", Math.round(maxValue));
        }
        if (accountBalance < amount) {
            return "درج کردہ رقم آپ کے موجودہ بیلنس سے زیادہ ہے";
        }
        return null;
    }


    //-------------------------------------//
    //----- Getters of the observers ------//
    //-------------------------------------//

    public LiveData<PersonalInfoData> getDriverProfile() {
        if (driverProfile == null) {
            driverProfile = new MutableLiveData<>();
            loadUserProfile();
        }
        return driverProfile;
    }

    public MutableLiveData<Boolean> getShowConfirmationDialog() {
        if (_showConfirmationDialog == null) {
            _showConfirmationDialog = new MutableLiveData<>();
        }
        return _showConfirmationDialog;
    }

    public MutableLiveData<Boolean> getShowLoader() {
        if (_showLoader == null) {
            _showLoader = new MutableLiveData<>();
        }
        return _showLoader;
    }

    public MutableLiveData<Boolean> getIsWithdrawCompleted() {
        if (_onWithdrawCompleted == null) {
            _onWithdrawCompleted = new MutableLiveData<>();
        }
        return _onWithdrawCompleted;
    }

    public MutableLiveData<String> getErrorMessage() {
        if (_errorMessage == null) {
            _errorMessage = new MutableLiveData<>();
        }
        return _errorMessage;
    }

    public LiveData<List<WithdrawPaymentMethod>> getAvailablePaymentMethods() {
        if (mAvailablePaymentMethods == null) {
            mAvailablePaymentMethods = new MutableLiveData<>();
            loadWithdrawalMethods();
        }
        return mAvailablePaymentMethods;
    }

    public WithdrawPaymentMethod getSelectedPaymentMethod() {
        return selectedPaymentMethod;
    }

    public MutableLiveData<Integer> getBalanceToWithdraw() {
        if (_balanceInt == null) {
            _balanceInt = new MutableLiveData<>();
        }
        return _balanceInt;
    }

    public void setSelectedPaymentMethod(WithdrawPaymentMethod selectedPaymentMethod) {
        this.selectedPaymentMethod = selectedPaymentMethod;
    }
}

File: app/src/main/java/com/bykea/pk/partner/ui/withdraw/WithdrawalViewModel.kt
package com.bykea.pk.partner.ui.withdraw

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.bykea.pk.partner.DriverApp
import com.bykea.pk.partner.R
import com.bykea.pk.partner.dal.source.remote.data.PersonalInfoData
import com.bykea.pk.partner.dal.source.remote.data.WithdrawPaymentMethod
import com.bykea.pk.partner.dal.source.withdraw.WithdrawRepository
import com.bykea.pk.partner.ui.helpers.AppPreferences
import com.bykea.pk.partner.utils.Constants
import org.apache.commons.lang3.StringUtils

/**
 * This is the view model class of {WithdrawalActivity}
 *
 * @author Arsal Imam
 */
class WithdrawalViewModel
/**
 * Constructor of this viewModel
 * @param withdrawRepository instance to get and update data
 */
(private val withdrawRepository: WithdrawRepository) : ViewModel() {

    /**
     * list of payment methods
     */
    private val _paymentMethods = MutableLiveData<List<WithdrawPaymentMethod>>().apply { value = emptyList() }
    val paymentMethods: LiveData<List<WithdrawPaymentMethod>>
        get() = _paymentMethods

    /**
     * observable to show/hide confirmation dialog
     */
    private val _showConfirmationDialog = MutableLiveData<Boolean>().apply { value = false }
    val showConfirmationDialog: LiveData<Boolean>
        get() = _showConfirmationDialog

    /**
     * callback event to view on withdrawal process completed
     */
    private val _onWithdrawCompleted = MutableLiveData<Boolean>().apply { value = false }
    val onWithdrawCompleted: LiveData<Boolean>
        get() = _onWithdrawCompleted

    /**
     * observable to show/hide loader
     */
    private val _showLoader = MutableLiveData<Boolean>().apply { value = false }
    val showLoader: LiveData<Boolean>
        get() = _showLoader

    /**
     * balance entered by user
     */
    private val _balanceInt = MutableLiveData<Int>().apply { value = 0 }
    val balanceInt: LiveData<Int>
        get() = _balanceInt

    /**
     * validation message to show on view
     */
    private val _errorMessage = MutableLiveData<String>().apply { value = null }
    val errorMessage: LiveData<String>
        get() = _errorMessage

    /**
     * driver profile obtained from API {@link #loadUserProfile()}
     */
    private val _driverProfile = MutableLiveData<PersonalInfoData>().apply { value = null }
    val driverProfile: LiveData<PersonalInfoData>
        get() = _driverProfile

    /**
     * observable for selected payment method
     */
    private var _selectedPaymentMethod: WithdrawPaymentMethod? = null
    var selectedPaymentMethod: WithdrawPaymentMethod? = null
        get() = _selectedPaymentMethod

    /**
     * Load user (driver) profile from repository
     */
    fun loadUserProfile() {
        _showLoader?.value = true
        withdrawRepository.getDriverProfile(object : WithdrawRepository.LoadWithdrawalCallback<PersonalInfoData?> {
            override fun onDataLoaded(data: PersonalInfoData?) {
                _driverProfile.value = data
                loadWithdrawalMethods()
            }

            override fun onDataNotAvailable(errorMsg: String) {
                _showLoader?.value = false
            }
        })
    }

    /**
     * returns driver's CNIC number if profile loaded
     *
     * @return cnic number
     */
    fun getDriverCnicNumber(): String {
        return if (driverProfile != null && driverProfile.value != null) {
            driverProfile.value!!.cnic
        } else {
            StringUtils.EMPTY
        }
    }

    /**
     * Load all payment methods from repository
     */
    fun loadWithdrawalMethods() {
        withdrawRepository.getAllPaymentMethods(object : WithdrawRepository.LoadWithdrawalCallback<List<WithdrawPaymentMethod>> {
            override fun onDataLoaded(data: List<WithdrawPaymentMethod>?) {
                if (!data.isNullOrEmpty()) {
                    data[0].isSelected = true
                    _selectedPaymentMethod = data[0]
                    _paymentMethods?.value = data
                }
                _showLoader?.value = false
            }

            override fun onDataNotAvailable(errorMsg: String) {
                _showLoader?.value = false
                _errorMessage?.value = DriverApp.getContext().getString(R.string.something_went_wrong)
            }
        })
    }

    /**
     * Executes withdrawal PUT API from repository to perform withdrawal operation.
     */
    fun confirmWithdraw() {
        val enteredBalance = balanceInt?.value!!
        if (enteredBalance != null) {

            val amount = balanceInt?.value!!
            _showLoader?.value = true

            withdrawRepository.performWithdraw(amount, selectedPaymentMethod?.code!!, object : WithdrawRepository.LoadWithdrawalCallback<Boolean> {

                override fun onDataLoaded(data: Boolean?) {
                    _showLoader?.value = false
                    _onWithdrawCompleted?.value = true
                }

                override fun onDataNotAvailable(errorMsg: String) {
                    _showLoader?.value = false
                    if (errorMsg.contentEquals(Constants.ApiError.ERROR_MSG_CODE.toString())) {
                        _errorMessage?.value = DriverApp.getContext().getString(R.string.withdraw_error_threshold_exceed)
                    } else {
                        _errorMessage?.value = DriverApp.getContext().getString(R.string.something_went_wrong)
                    }
                }
            })
        }
    }

    /**
     * Binded with button's onclick,
     * @param amt amount entered by the user
     */
    fun onSubmitClicked(amt: String?) {
        try {
            val amount = Integer.valueOf(amt)
            val errorMsg = validateContent(amount)
            _errorMessage?.value = errorMsg
            if (errorMsg == null) {
                _balanceInt?.value = amount
                _showConfirmationDialog?.value = true
            }
        } catch (e: NumberFormatException) {
            e.printStackTrace()
        }

    }

    /**
     * this method will show/hide the confirmation dialog
     *
     * @param isShow show(true) / hide(false)
     */
    fun showConfirmationDialog(isShow: Boolean) {
        _showConfirmationDialog.value = isShow
    }

    /**
     * Performing validation on form submit
     * @param amount to withdraw
     *
     * @return **null** if validation done successfully, otherwise returns error message
     */
    private fun validateContent(amount: Int): String? {
        try {
            val settingsData = AppPreferences.getSettings()

            val maxValue = settingsData.settings.withdrawPartnerMaxLimit
            val minValue = settingsData.settings.withdrawPartnerMinLimit

            val accountBalance = driverProfile?.value?.wallet!!

            if (accountBalance < amount) {
                return DriverApp.getContext().getString(R.string.dont_have_enough_to_withdraw)
            }
            if (amount < minValue) {
                return String.format(
                        DriverApp.getContext().getString(R.string.minimum_amount_to_withdraw_2)
                                + StringUtils.SPACE
                                + DriverApp.getContext().getString(R.string.formatted_price)
                                + StringUtils.SPACE
                                + DriverApp.getContext().getString(R.string.minimum_amount_to_withdraw_1), Math.round(minValue))
            }
            if (amount > maxValue) {
                return String.format(DriverApp.getContext().getString(R.string.maximum_amount_to_withdraw_1)
                        + StringUtils.SPACE
                        + DriverApp.getContext().getString(R.string.formatted_price)
                        + StringUtils.SPACE
                        + DriverApp.getContext().getString(R.string.maximum_amount_to_withdraw_2), Math.round(maxValue))
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return DriverApp.getContext().getString(R.string.something_went_wrong)
        }
        return null
    }

    /**
     * this method will hide the validation layout
     */
    fun removeWarnings() {
        _errorMessage.value = null
    }
}